DSL for programmatic test plan generation

vlsi opened the pull request:
## Description

See https://lists.apache.org/thread/52o722d7npsjj99m4k2zdz44mstl02lg

## TODO

* [ ] Should there be an intermediate test plan representation or should the current `ListedHashTree` and other classes represent the tree? I incline we should create a separate set of classes to avoid leaking unwanted methods to the end-user. For instance, methods like `#sample()`, `#interrupt()` do not seem to make sense during test plan generation.
* [ ] Should DSL functions return objects or should they be `no-result` functions?
* [ ] How should `$` be escaped? (see https://youtrack.jetbrains.com/issue/KT-2425, https://lists.apache.org/thread/717sm4k8kx6ojmkq0j1w0fhkv41bkm8d)
* [ ] Should the DSL incline to name-less elements or name-full elements? JMeter GUI requires names for all elements, including `if` and `pauses`, however, regular programming languages and DSLs do not require names for things like `if`, `for`, etc. Of course, the transaction controller should have a name, but is the name required for `httpSampler`?
* [ ] Should there be `.children()` or something like that API to access all the children elements of the currently built node?

## Motivation and Context

(in alphabetical order)
https://aliesbelik.github.io/awesome-jmeter/#dsl
https://github.com/abstracta/jmeter-java-dsl
https://github.com/anasoid/jmeter-as-code
https://github.com/smicyk/groovy-jmeter

https://github.com/anasoid/jmeter-as-code/issues/133

https://github.com/Kotlin/kotlinx.html/wiki/Micro-templating-and-DSL-customizing

## Screenshots (if appropriate):

<img width="408" alt="Copy Code action in JMeter UI" src="https://github.com/apache/jmeter/assets/213894/fa48982b-aeeb-42d9-917d-9a6fbb96b9b3">

```kotlin
org.apache.jmeter.threads.ThreadGroup::class {
    props {
        it[guiClass] = "org.apache.jmeter.threads.gui.ThreadGroupGui"
        it[name] = "Thread Group"
        it[mainController] = org.apache.jmeter.control.LoopController().apply {
            props {
                it[name] = "Loop Controller"
                it[guiClass] = "org.apache.jmeter.control.gui.LoopControlPanel"
                it[testClass] = "org.apache.jmeter.control.LoopController"
                it[loops] = "1"
            }
        }
        it[numThreads] = "1"
        it[rampTime] = "1"
    }

    org.apache.jmeter.protocol.http.sampler.HTTPSamplerProxy::class {
        props {
            it[arguments] = org.apache.jmeter.config.Arguments().apply {
                props {
                    it[arguments] = listOf(
                        org.apache.jmeter.protocol.http.util.HTTPArgument().apply {
                            props {
                                it[value] = "World"
                                it[metadata] = "="
                                it[useEquals] = true
                                it[argumentName] = "user"
                            }
                        },
                        org.apache.jmeter.protocol.http.util.HTTPArgument().apply {
                            props {
                                it[alwaysEncode] = true
                                it[value] = "asdfa"
                                it[metadata] = "="
                                it[useEquals] = true
                                it[argumentName] = "test"
                            }
                        },
                    )
                    it[name] = "User Defined Variables"
                    it[guiClass] = "org.apache.jmeter.protocol.http.gui.HTTPArgumentsPanel"
                    it[testClass] = "org.apache.jmeter.config.Arguments"
                }
            }
            it[domain] = "example.com"
            it[path] = "/api/v1/hello"
            it[method] = "GET"
            it[followRedirects] = true
            it[useKeepalive] = true
            it[proxy.scheme] = "https"
            it[proxy.host] = "localhost"
            it[proxy.port] = "8080"
            it[proxy.username] = "secret"
            it[proxy.password] = "password1"
            it[name] = "HTTP Request"
            it[guiClass] = "org.apache.jmeter.protocol.http.control.gui.HttpTestSampleGui"
        }

        org.apache.jmeter.extractor.RegexExtractor::class {
            props {
                it[guiClass] = "org.apache.jmeter.extractor.gui.RegexExtractorGui"
                it[name] = "Regular Expression Extractor"
                it[referenceName] = "regexVar"
                it[regularExpression] = "hello\\s+?world"
                it[template] = "\$1\$"
            }
        }

        org.apache.jmeter.protocol.http.control.HeaderManager::class {
            props {
                it[headers] = listOf(
                    org.apache.jmeter.protocol.http.control.Header().apply {
                        props {
                            it[headerName] = "Accept"
                            it[value] = "text/plain"
                        }
                    },
                    org.apache.jmeter.protocol.http.control.Header().apply {
                        props {
                            it[headerName] = "User-Agent"
                            it[value] = "JMeter"
                        }
                    },
                    org.apache.jmeter.protocol.http.control.Header().apply {
                        props {
                            it[headerName] = "X-JMeter-Thread"
                            it[value] = "Thread \${__threadNum}"
                        }
                    },
                )
                it[guiClass] = "org.apache.jmeter.protocol.http.gui.HeaderPanel"
                it[name] = "HTTP Header Manager"
            }
        }
    }

    org.apache.jmeter.control.LoopController::class {
        props {
            it[guiClass] = "org.apache.jmeter.control.gui.LoopControlPanel"
            it[name] = "Loop Controller"
            it[loops] = "1"
        }

        org.apache.jmeter.sampler.DebugSampler::class {
            props {
                it[guiClass] = "org.apache.jmeter.testbeans.gui.TestBeanGUI"
                it[name] = "Debug Sampler"
            }
        }
    }
}
```

kirillyu commented:
Initially - thanks for the idea, this is really what the community needs.
 
My opinion on current issues:
1. I am inclined more to the idea that an intermediate presentation is needed than not.  
2. Can you give details on this?  I don't see any clear advantage in using no-result functions.  
3. This is an interesting point.  In my opinion there is no need to escape this character.  It should also be used as part of the string, and replace will happen in the jmeter engine.  In other cases, the variables must be taken from the context in the vars.get style.  
4. name-less with a description is better, it is faster and takes up less space.  
5. Yes, using .children () is more convenient for implementing custom builders.

vlsi committed:
feat: add DSL for programmatic test plan creation

vlsi commented:
I've added documentation, and a few more samples

vlsi committed:
test: add test that verifies if thread groups interrupt the running threads

vlsi committed:
Open Model Thread Group: interrupt pending HTTP requests and other Interruptible test elements on test stop

Previously, OMTG was just issuing java.lang.Thread.interrupt(),
however certain samplers require special treatment.

So now OMTG calls JMeterThread.interrupt() as well, so Interruptible
elements are interrupted as well.

vlsi committed:
test: add test for replacing variables in Include Controller

See https://github.com/apache/jmeter/pull/679

vlsi committed:
Add short names to the properties

vlsi committed:
feat: add Copy Code action to copy test element as code

It produces a draft code, so it is easier to start using DSL
by generating code from the UI.

vlsi commented:
I've added `Copy Code` action so users could copy a code draft:
<img width="408" alt="Copy Code action in JMeter UI" src="https://github.com/apache/jmeter/assets/213894/fa48982b-aeeb-42d9-917d-9a6fbb96b9b3">

Genenerates the following:

```kotlin
org.apache.jmeter.threads.ThreadGroup::class {
    props {
        it[guiClass] = "org.apache.jmeter.threads.gui.ThreadGroupGui"
        it[testClass] = "org.apache.jmeter.threads.ThreadGroup"
        it[name] = "Thread Group"
        it[mainController] = org.apache.jmeter.control.LoopController().apply {
            props {
                it[name] = "Loop Controller"
                it[guiClass] = "org.apache.jmeter.control.gui.LoopControlPanel"
                it[testClass] = "org.apache.jmeter.control.LoopController"
                it[comments] = ""
            }
            setProperty("LoopController.continue_forever", false)
            setProperty("LoopController.loops", "1")
        }
        it[numThreads] = "1"
        it[rampTime] = "1"
        it[duration] = ""
        it[delay] = ""
        it[comments] = ""
    }
    setProperty("ThreadGroup.on_sample_error", "continue")

    org.apache.jmeter.protocol.http.sampler.HTTPSamplerProxy::class {
        props {
            it[followRedirects] = true
            it[useKeepalive] = true
            it[name] = "HTTP Request"
            it[guiClass] = "org.apache.jmeter.protocol.http.control.gui.HttpTestSampleGui"
            it[testClass] = "org.apache.jmeter.protocol.http.sampler.HTTPSamplerProxy"
            it[comments] = ""
        }
        setProperty("HTTPsampler.Arguments", org.apache.jmeter.config.Arguments().apply {
            props {
                it[name] = "User Defined Variables"
                it[guiClass] = "org.apache.jmeter.protocol.http.gui.HTTPArgumentsPanel"
                it[testClass] = "org.apache.jmeter.config.Arguments"
                it[comments] = ""
            }
            setProperty("Arguments.arguments", listOf(
                org.apache.jmeter.protocol.http.util.HTTPArgument().apply {
                    setProperty("HTTPArgument.always_encode", false)
                    setProperty("Argument.value", "world")
                    setProperty("Argument.metadata", "=")
                    setProperty("HTTPArgument.use_equals", true)
                    setProperty("Argument.name", "hello")
                },
                org.apache.jmeter.protocol.http.util.HTTPArgument().apply {
                    setProperty("HTTPArgument.always_encode", true)
                    setProperty("Argument.value", "asdfa")
                    setProperty("Argument.metadata", "=")
                    setProperty("HTTPArgument.use_equals", true)
                    setProperty("Argument.name", "test")
                },
            ))
        })
        setProperty("HTTPSampler.domain", "")
        setProperty("HTTPSampler.port", "")
        setProperty("HTTPSampler.protocol", "")
        setProperty("HTTPSampler.contentEncoding", "")
        setProperty("HTTPSampler.path", "")
        setProperty("HTTPSampler.method", "GET")
        setProperty("HTTPSampler.auto_redirects", false)
        setProperty("HTTPSampler.embedded_url_re", "")
        setProperty("HTTPSampler.connect_timeout", "")
        setProperty("HTTPSampler.response_timeout", "")

        org.apache.jmeter.extractor.RegexExtractor::class {
            props {
                it[guiClass] = "org.apache.jmeter.extractor.gui.RegexExtractorGui"
                it[testClass] = "org.apache.jmeter.extractor.RegexExtractor"
                it[name] = "Regular Expression Extractor"
                it[comments] = ""
            }
            setProperty("RegexExtractor.useHeaders", "false")
            setProperty("RegexExtractor.refname", "asdf")
            setProperty("RegexExtractor.regex", "asdf")
            setProperty("RegexExtractor.template", "asdf")
            setProperty("RegexExtractor.default", "")
            setProperty("RegexExtractor.match_number", "")
        }
    }

    org.apache.jmeter.control.LoopController::class {
        props {
            it[guiClass] = "org.apache.jmeter.control.gui.LoopControlPanel"
            it[testClass] = "org.apache.jmeter.control.LoopController"
            it[name] = "Loop Controller"
            it[comments] = ""
        }
        setProperty("LoopController.continue_forever", true)
        setProperty("LoopController.loops", "1")
    }
}
```

vlsi committed:
chore: use schema in AbstractTestElement for properties

vlsi committed:
chore: use TestElement.setEnabled instead of using the property name

vlsi committed:
feat: add GenericControllerSchema, and LoopControllerSchema

vlsi committed:
feat: add HTTPArgumentSchema

vlsi committed:
feat: add TestElement.remove(PropertyDescriptor), TestElement.getOrNull(..)

vlsi committed:
feat: add HttpSamplerBaseSchema

vlsi committed:
feat: add ability to factor properties into subgroups

It enables creating a subgroup and attach it to a schema or to another group.

vlsi committed:
feat: Add AbstractThreadGroupSchema.onSampleError

vlsi committed:
feat: add http.HeaderSchema

vlsi committed:
feat: add RegexExtractorSchema

vlsi committed:
feat: add DebugSamplerSchema

vlsi committed:
feat: add HeaderManagerSchema

vlsi committed:
feat: add WhileControllerSchema

vlsi committed:
feat: add TransactionControllerSchema

vlsi committed:
feat: add ArgumentSchema, ArgumentsSchema

vlsi committed:
chore: use abstract classes for schemas

A single instance for the schema is enough, so it would be better
using abstract class so the users know they don't need instantiate it.

vlsi committed:
feat: add IfControllerSchema

vlsi committed:
doc: explain the reason for using org.apache.jmeter.treebuilder.Action interface

Java's interfaces are bad since they have no declaration side type variance.
Kotlin's ... -> Unit is hard to use in Java as it would require "return Unit.INSTANCE"

Regular Kotlin's functions A->B are fine though.
